SourceCode/src/Ray.cpp

```diff
@@ -1,63 +1,121 @@
 #include <tracer/Ray.h>
 
+// #include <limits>
 #include <cmath>
 #include <limits>
 #include <optional>
 
 #include "tracer/Triangle.h"
 #include "tracer/Vector.h"
 
-Ray::Ray() = default;
 
-Ray::Ray(const Vector& origin, const Vector& direction) : origin(origin), direction(direction){};
 
-std::optional<Vector> Ray::intersectWithTriangle(const Triangle& triangle) const {
+#if 0
+
+std::optional<Vector> Ray::intersectWithTriangle(const Triangle& triangle, const Vector& cameraPosition) const {
   Vector triangleNormal = triangle.getTriangleNormal();
   float normalDotRayDirection = this->direction.dot(triangleNormal);
 
+  //   if (normalDotRayDirection >= 0) {
+  //     return {};
+  //   }
+
   // if N . R ~= 0, the ray is parallel to the plane - no intersection or too far away
-  if (std::fabs(normalDotRayDirection) < std::numeric_limits<float>::epsilon()) {
+  // if (std::fabs(normalDotRayDirection) < std::numeric_limits<float>::epsilon()) {
+  //   return {};
+  // }
+
+  float distanceToPlane = (triangle[0] - cameraPosition).dot(triangleNormal);
+
+  float t = distanceToPlane / normalDotRayDirection;
+
+  Vector intersectionPoint = this->origin + t * this->direction;
+
+  if (!triangle.pointIsInTriangle(intersectionPoint)) {
     return {};
   }
 
-  // TODO(daniel):
-  float distanceToPlane = -(triangle[0] - this->origin).dot(triangleNormal);
+  return intersectionPoint;
+}
 
-  float t = -(this->origin.dot(triangleNormal) + distanceToPlane) / (normalDotRayDirection);
+std::optional<Vector> ShadowRay::intersectWithTriangle(const Triangle& triangle, const Vector& cameraPosition) const {
+  Vector triangleNormal = triangle.getTriangleNormal();
+  float normalDotRayDirection = this->direction.dot(triangleNormal);
 
-  // if t < 0, the triangle is facing in the wrong direction
-  if (t < 0) {
+  if (normalDotRayDirection == 0) {
     return {};
   }
 
+  // if N . R ~= 0, the ray is parallel to the plane - no intersection or too far away
+  // if (std::fabs(normalDotRayDirection) < std::numeric_limits<float>::epsilon()) {
+  //   return {};
+  // }
+
+  float distanceToPlane = (triangle[0] - cameraPosition).dot(triangleNormal);
+
+  float t = distanceToPlane / normalDotRayDirection;
+
   Vector intersectionPoint = this->origin + t * this->direction;
 
-  Vector e0 = triangle[1] - triangle[0];
-  Vector c0 = intersectionPoint - triangle[0];
-  if (triangleNormal.dot(e0 * c0) < 0) {
+  if (!triangle.pointIsInTriangle(intersectionPoint)) {
+    return {};
+  }
+
+  return intersectionPoint;
+}
+
+#else
+template <>
```

Comment:
You're repeating the code 3 times, consider using ray flags instead of templates